### PR TITLE
Rakefile, ignores, defines and errors

### PR DESCRIPTION
**Pull request by mattbenic**
-Windows build in rakefile ensures VS tools are on path, and prevents ATOMIC_DEV_BUILD from being set so resource paths work correctly
-Cache and compiled shaders excluded from git
-Typescript definitions updated from latest build-these should actually probably be ignored instead
-Added a typescript definitions file for items not automatically generated by JSBind
-More detail added to errors logged when loading plugins
-Debug to the debugger console when debugger is attached

### CODE REVIEW

File: Script/TypeScript/AtomicExtended.d.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+// Manual typescript definitions from various CPP and JS files that can't yet be autogenerated
+// no-default-lib and modified lib.d.ts used to allow redefinition of self
+/// <reference no-default-lib="true"/>
+/// <reference path="./lib.d.ts" />
```
Comment (JoshEngebretson) [inline Script/TypeScript/AtomicExtended.d.ts]: Does this require a custom lib.d.ts?  I am not seeing it here?

File: Script/TypeScript/AtomicExtended.d.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+declare function getRenderer(): Atomic.Renderer;
+declare function getResourceCache(): Atomic.ResourceCache;
+declare function getInput(): Atomic.Input;
+declare function getFileSystem(): Atomic.FileSystem;
+declare function getNetwork(): Atomic.Network;
+declare function script(script: string): boolean;
+declare function destroy(node: Atomic.Node): boolean;
+declare function destroy(scene: Atomic.Scene): boolean;
+declare function destroy(component: Atomic.JSComponent): boolean;
+
```
Comment (JoshEngebretson) [inline Script/TypeScript/AtomicExtended.d.ts]: I think there is some overlap with https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Script/TypeScript/AtomicWork.d.ts

I called it AtomicWork as at some point want to automatically merge the stuff in here with Atomic.d.ts

It also looks like some of these are declared as global functions, in AtomicWork they are declared as exports from the Atomic module:

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Script/TypeScript/AtomicWork.d.ts

File: Script/TypeScript/AtomicExtended.d.ts (hunk 1 above)
Comment (mattbenic) [inline Script/TypeScript/AtomicExtended.d.ts]: Sorry I just removed that, missed the ref. Will nuke it now push


Regards,
Matt Benic
mattbenic@gmail.com | +27 (72) 1826161 | za.linkedin.com/in/mattbenic

File: Script/TypeScript/AtomicExtended.d.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+// From JSRequire.cpp
+declare function require(path: string): any;
```
Comment (JoshEngebretson) [inline Script/TypeScript/AtomicExtended.d.ts]: Require is part of the CommonJS spec, JSRequire handles the search path:

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicJS/Javascript/JSRequire.cpp#L185

File: Script/TypeScript/AtomicExtended.d.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        getSpriteSheet2D(xmlFile: string): Resource;
+        getSpriteSheet(xmlFile: string): Resource;
+        getSound(soundFile: string): Resource;
+        getSprite2D(spriteFile: string): Resource;
+        showDebugHud();
+        createScene2D(): Scene;
+        dumpMetrics();
+        createScene3D(filename: string): Scene;
+    }
+    var game: Game;
```
Comment (JoshEngebretson) [inline Script/TypeScript/AtomicExtended.d.ts]: Atomic game won't be provided in core anymore, it is really just some convenience stuff that people can include in there game as a module.  Atomic module itself now holds shortcuts to graphics, input, etc subsystems

File: Script/TypeScript/AtomicExtended.d.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        EVENT_TYPE_TAB_CHANGED: number;
+        EVENT_TYPE_CUSTOM: number;
+        VISIBILITY_GONE: number;
+        VISIBILITY_INVISIBLE: number;
+        VISIBILITY_VISIBLE: number;
+
+        debugGetWrappedWidgetCount(): number;
+        debugGetUIKeepAliveCount(): number;
+        debugShowSettingsWindow(): number;
+    };
```
Comment (JoshEngebretson) [inline Script/TypeScript/AtomicExtended.d.ts]: This UI was all interim, glad this reminds me to remove it.  It is all now automatically generated:

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Script/TypeScript/Atomic.d.ts#L578

:)

File: Rakefile [hunk 6]
```diff
@@ -58,8 +58,12 @@ namespace :build  do
     end
 
     Dir.chdir(CMAKE_WINDOWS_BUILD_FOLDER) do
+      # ensure VS commandline tools are on the path
+      VSCOMNTOOLS = ENV['VS140COMNTOOLS']
 
-      sh "cmake ../../ -G \"Visual Studio 14 2015\""
+      sh "\"#{VSCOMNTOOLS}vsvars32.bat\""
+
+      sh "cmake ../../ -DATOMIC_DEV_BUILD=0 -G \"Visual Studio 14 2015\""
```
Comment (JoshEngebretson) [inline Rakefile]: There is a variable for this:

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/30b5ea42ecd21443d39792a273770c6ff8560de9/Rakefile#L17

Which gets cleared when packaging

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/30b5ea42ecd21443d39792a273770c6ff8560de9/Rakefile#L207

It is used here:

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/30b5ea42ecd21443d39792a273770c6ff8560de9/Rakefile#L79

Looks like it needs to be added to Windows, I have been running from the packaged Artifacts folder, will test that can move it once the dev build is set properly

File: Script/TypeScript/Atomic.d.ts [hunk 7]
```diff
@@ -318,6 +318,30 @@ declare module Atomic {
    export var RAY_TRIANGLE_UV: RayQueryLevel;
 
 
+   // enum RenderCommandType
```
Comment (JoshEngebretson) [inline Script/TypeScript/Atomic.d.ts]: The thinking on Atomic.d.ts being in the repo is that you would be able to use it to compile without building the engine, though I think that should probably be ignored too.  I am thinking more of packaging node with the repo which is probably pretty easy to do

File: Source/Atomic/Core/Object.h [hunk 8]
```diff
@@ -40,8 +40,8 @@ class EventHandler;
         virtual const Atomic::String& GetTypeName() const { return GetTypeNameStatic(); } \
         static Atomic::StringHash GetTypeStatic() { static const Atomic::StringHash typeStatic(#typeName); return typeStatic; } \
         static const Atomic::String& GetTypeNameStatic() { static const Atomic::String typeNameStatic(#typeName); return typeNameStatic; } \
-        virtual ClassID GetClassID() const { return GetClassIDStatic(); } \
-        static ClassID GetClassIDStatic() { static const int typeID = 0; return (ClassID) &typeID; }
+        virtual Atomic::ClassID GetClassID() const { return GetClassIDStatic(); } \
+        static Atomic::ClassID GetClassIDStatic() { static const int typeID = 0; return (Atomic::ClassID) &typeID; }
```
Comment (JoshEngebretson) [inline Source/Atomic/Core/Object.h]: Good catch

File: .gitignore [hunk 9]
```diff
@@ -1,4 +1,6 @@
 
+*.ps3
+*.vs3
```
Comment (JoshEngebretson) [inline .gitignore]: Thanks, been meaning to do this!

File: Rakefile [hunk 10]
```diff
@@ -58,8 +58,12 @@ namespace :build  do
     end
 
     Dir.chdir(CMAKE_WINDOWS_BUILD_FOLDER) do
+      # ensure VS commandline tools are on the path
+      VSCOMNTOOLS = ENV['VS140COMNTOOLS']
 
-      sh "cmake ../../ -G \"Visual Studio 14 2015\""
+      sh "\"#{VSCOMNTOOLS}vsvars32.bat\""
```
Comment (JoshEngebretson) [inline Rakefile]: I think this is a good add, didn't break the CI here which is a good sign :)

File: Script/TypeScript/AtomicExtended.d.ts (hunk 2 above)
Comment (mattbenic) [inline Script/TypeScript/AtomicExtended.d.ts]: There are both get accessors and properties for subsystems (eg Atomic.getGraphics and Atomic.graphics). I see the accessors haven't been added to AtomicWork.d.ts, is this intentional or can I add them.

**Comment by JoshEngebretson** [comment]
Great, landing!

**Pull request merged**